adapters/sovrn/sovrn.go

```diff
@@ -0,0 +1,274 @@
+package sovrn
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"github.com/prebid/prebid-server/pbs"
+	"golang.org/x/net/context/ctxhttp"
+)
+
+type SovrnAdapter struct {
+	http *adapters.HTTPAdapter
+	URI  string
+}
+
+// Name - export adapter name */
+func (s *SovrnAdapter) Name() string {
+	return "sovrn"
+}
+
+// FamilyName used for cookies and such
+func (s *SovrnAdapter) FamilyName() string {
+	return "sovrn"
+}
+
+type sovrnParams struct {
+	TagID int `json:"tagid"`
+}
+
+func (s *SovrnAdapter) SkipNoCookies() bool {
+	return false
+}
+
+// Call send bid requests to sovrn and receive responses
+func (s *SovrnAdapter) Call(ctx context.Context, req *pbs.PBSRequest, bidder *pbs.PBSBidder) (pbs.PBSBidSlice, error) {
+	supportedMediaTypes := []pbs.MediaType{pbs.MEDIA_TYPE_BANNER}
+	sReq, err := adapters.MakeOpenRTBGeneric(req, bidder, s.FamilyName(), supportedMediaTypes, true)
+
+	if err != nil {
+		return nil, err
+	}
+
+	sovrnReq := openrtb.BidRequest{
+		ID:   sReq.ID,
+		Imp:  sReq.Imp,
+		Site: sReq.Site,
+	}
+
+	// add tag ids to impressions
+	for i, unit := range bidder.AdUnits {
+		var params sovrnParams
+		err = json.Unmarshal(unit.Params, &params)
+		if err != nil {
+			return nil, err
+		}
+		sovrnReq.Imp[i].Secure = sReq.Imp[i].Secure
+		sovrnReq.Imp[i].TagID = strconv.Itoa(params.TagID)
+		sovrnReq.Imp[i].Banner.Format = nil // todo: need to do this?
+	}
+
+	reqJSON, err := json.Marshal(sovrnReq)
+	if err != nil {
+		return nil, err
+	}
+
+	debug := &pbs.BidderDebug{
+		RequestURI: s.URI,
+	}
+
+	if req.IsDebug {
+		debug.RequestBody = string(reqJSON)
+		bidder.Debug = append(bidder.Debug, debug)
+	}
+
+	httpReq, _ := http.NewRequest("POST", s.URI, bytes.NewReader(reqJSON))
+	httpReq.Header.Set("Content-Type", "application/json")
+	httpReq.Header.Set("User-Agent", sReq.Device.UA)
```

Comment:
This code looks like it'll `panic` if `Device` or `User` is `nil`. You'll need to handle those cases more gracefully somehow.

Please add tests for them too.

Reply:
Will it be possible for those fields to be `nil`? These fields should be populated from the prebid.js request, no?

Reply:
It doesn't really matter. These adapters are run in their own goroutines for parallelism, so any `panic` code causes the process to segfault + crash.

It's a major vulnerability to PBS hosts, since anyone can just `curl` or `Postman` a request in.

Servers should never trust client input.

Reply:
Fair enough, I'll rework this.